subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/repositories/resolver/ExternalResourceResolverDescriptorParseContext.java

```diff
@@ -49,10 +60,25 @@ public ExternalResourceResolverDescriptorParseContext(ComponentResolvers mainRes
     }
 
     public LocallyAvailableExternalResource getMetaDataArtifact(ModuleComponentIdentifier moduleComponentIdentifier, ArtifactType artifactType) {
+        moduleComponentIdentifier = resolveDynamicVersionIfNecessary(moduleComponentIdentifier);
         File resolvedArtifactFile = resolveMetaDataArtifactFile(moduleComponentIdentifier, mainResolvers.getComponentResolver(), mainResolvers.getArtifactResolver(), artifactType);
         return fileResourceRepository.resource(resolvedArtifactFile);
     }
 
+    private ModuleComponentIdentifier resolveDynamicVersionIfNecessary(ModuleComponentIdentifier moduleComponentIdentifier) {
+        ModuleVersionSelector selector = DefaultModuleVersionSelector.newSelector(moduleComponentIdentifier.getGroup(),
+            moduleComponentIdentifier.getModule(),
+            moduleComponentIdentifier.getVersion());
+        DependencyMetadata metadata = new MavenDependencyMetadata(MavenScope.Compile, false, selector, new ArrayList<Artifact>(), new ArrayList<Exclude>());
```

Comment:
This type `(ExternalResourceResolverDescriptorParseContext`) should not be tied to Maven in this way. The simplest option is probably to have  the caller supply a `DependencyMetadata` instance instead of a `ModuleComponentSelector` (when we need to support dynamic versions). 

So my previous suggestion becomes:
```
public interface DescriptorParseContext {
    LocallyAvailableExternalResource getMetaDataArtifact(ModuleComponentIdentifier componentIdentifier, ArtifactType artifactType);
    LocallyAvailableExternalResource getMetaDataArtifact(DependencyMetadata dependency, ArtifactType artifactType);
}
```

Reply:
Done. Please take a look at it again.